### DISCUSSION TITLE
Method body macro-like templates

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6228] [Language Ideas]
I find a lot of repetitive code in many methods about loading data from a database, like
```csharp
try
{
    // Something
}
catch (Exception e)
{
    logger.LogException(e);
}
```

All that code being present in multiple methods per class, multiple classes across multiple projects. All this boilerplate adds up and reduces code quality, by giving the programmer less meaningful content in the code.

What I propose is having something like a macro preprocessor directive, specifically for method code templates:
```csharp
#macro method TryTemplate
try
{
    #body
}
catch (Exception e)
{
    logger.LogException(e);
}
#endmacro

public void LoadSomething()
#macro use TryTemplate
{
    // Do the loading here
}
```

The `LoadSomething()` method would then be the equivalent of:
```csharp
public void LoadSomething()
{
    try
    {
        // Do the loading here
    }
    catch (Exception e)
    {
        logger.LogException(e);
    }
}
```

*The syntax choices are premature, and subject to change, this is just a showcase of what the feature would pretty much look like.*

**Comment by DavidArno** [comment]
Since C# already supports lambdas, `Action` etc, I can already do this:
```cs
void TryTemplate(Action action)
{
    try
    {
        action();
    }
    catch (Exception e)
    {
        logger.LogException(e);
    }
}

void LoadSomething()
{
    TryTemplate(() => {
        // Do the loading here
    });
}
```

What advantage over this approach would macros offer us?

**Reply by foxesknow** [reply]
Performance?

No need to create a capture class if the lambda passed to `TryTemplate` captures local state?

**Reply by TahirAhmadov** [reply]
@foxesknow don't worry about the performance in this case - if TryTemplate is a static or non-virtual instance method, it's likely to be inlined anyway.

**Reply by Rekkonnect** [reply]
Consider a more complicated example:
```csharp
#macro method WithAnyCheck
bool any = !IsEmpty;
if (!any)
    return false;

#body
#endmacro

public bool HasOptimalValue()
#macro use WithAnyCheck
{
    // some random calculation
    int difference = values[^1] - values[0];
    return difference > values[Length / 2];
}
```

**Reply by TahirAhmadov** [reply]
@AlFasGD still, macros are not the solution. In your example, that can be added as a bool returning private method, and `if(this.WithAnyCheck()) return false;` where you need to use it. And again, it will be inlined most likely.

Macros have many problems, such as, how do you ensure compile-time type safety, etc.

**Reply by Rekkonnect** [reply]
It's more about reducing repetition, and the fact that you can never guarantee a method gets inlined gives you more control of the situation.

**Reply by TahirAhmadov** [reply]
There is an attribute you can use to hint that the method should be aggressively inlined. And yes, methods are there precisely to avoid repetition.

**Reply by Rekkonnect** [reply]
You still don't guarantee inlining, and in some cases inlining even reduces performance.

**Reply by HaloFour** [reply]
@AlFasGD 

> and in some cases inlining even reduces performance.

By preventing the method it is inlined into from being inlined itself, which is exactly what this macro language would also do.
